File: src/python_package/python_package/hw1.py
# This is the first node
# This file has 1 publisher
# Name: number_publisher
# This will publish under a topic named 'number'

import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from numpy.random import randint

class HW1(Node):
    def __init__(self):
        super().__init__("number_publisher")

        self.the_publisher_ = self.create_publisher(Int64, "number", 10)
        

        self.get_logger().info("number_publisher is created and published under number")

        self.create_timer(1.0, self.publish_message)

    def publish_message(self):
        random_number = randint(100)
        the_message = Int64()
        the_message.data = random_number
        self.get_logger().info(f"The random number is: {random_number}")
        self.the_publisher_.publish(the_message)



def main(args=None):
    rclpy.init(args=args)
    the_node = HW1()
    rclpy.spin(the_node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_package/src/hw1.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include <memory>
#include <chrono>
#include <random>

using namespace std::chrono_literals;

class NumberPublisher : public rclcpp::Node
{
public:
    NumberPublisher() : Node("number_publisher")
    {
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);

        RCLCPP_INFO(this->get_logger(), "number_publisher is created and published under number");

        timer_ = this->create_wall_timer(1s, std::bind(&NumberPublisher::publishMessage, this));
    }

private:
    void publishMessage()
    {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<int64_t> dis(0, 100);
        int64_t random_number = dis(gen);

        auto message = example_interfaces::msg::Int64();
        message.data = random_number;
        RCLCPP_INFO(this->get_logger(), "The random number is: %ld", random_number);

        publisher_->publish(message);
    }

    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
